scipy/special/specfun/specfun.f

```diff
@@ -5790,28 +5790,30 @@ SUBROUTINE HYGFZ(A,B,C,Z,ZHF,ISFER)
 C
         IMPLICIT DOUBLE PRECISION (A-H,O-Y)
         IMPLICIT COMPLEX *16 (Z)
-        LOGICAL L0,L1,L2,L3,L4,L5,L6
+        LOGICAL L0,L1,L2,L3,L4,L5,L6,L7
         X=DBLE(Z)
         Y=DIMAG(Z)
         EPS=1.0D-15
         ISFER=0
-        L0=C.EQ.INT(C).AND.C.LT.0.0D0
+        L0=C.EQ.INT(C).AND.C.LE.0.0D0
         L1=DABS(1.0D0-X).LT.EPS.AND.Y.EQ.0.0D0.AND.C-A-B.LE.0.0D0
         L2=CDABS(Z+1.0D0).LT.EPS.AND.DABS(C-A+B-1.0D0).LT.EPS
-        L3=A.EQ.INT(A).AND.A.LT.0.0D0
-        L4=B.EQ.INT(B).AND.B.LT.0.0D0
+        L3=A.EQ.INT(A).AND.A.LE.0.0D0
+        L4=B.EQ.INT(B).AND.B.LE.0.0D0
         L5=C-A.EQ.INT(C-A).AND.C-A.LE.0.0D0
         L6=C-B.EQ.INT(C-B).AND.C-B.LE.0.0D0
+        L7=(L3.AND.(DABS(A).LE.DABS(C)))
+     &       .OR.(L4.AND.(DABS(B).LE.DABS(C)))
```

Comment:
@FormerPhysicist, you write [my emphasis]:
> Now if z is complex and c=-m, this PR fixes hyp2f1 so that it returns the correctly computed value if and only if either a or b equals a negative integer -n where _**n<m**_.

Shouldn't it then be
```suggestion
        L7=(L3.AND.(DABS(A).LT.DABS(C)))
     &       .OR.(L4.AND.(DABS(B).LT.DABS(C)))
```

Reply:
@h-vetinari No, I think it should be LE (less than or equal).

When a=c (or b=c) and both are negative integers, then we need to let the method fall through and hit this section of [code](https://github.com/FormerPhysicist/scipy/blob/5837daef2fb8b8418b0ba97a4eee3583cc8ac5fc/scipy/special/specfun/specfun.f#L5832).

That part of the code implements formula 15.4.2 from Abramowitz and Stegun.

Otherwise, when a (or b) is a negative integer and `abs(c) > abs(a)`, then the hypergeometric function reduces to a polynomial that is analytic everywhere. And again in this case we need to let the method fall through and run the code [here](https://github.com/FormerPhysicist/scipy/blob/5837daef2fb8b8418b0ba97a4eee3583cc8ac5fc/scipy/special/specfun/specfun.f#L5840)

Note that the line you've highlighted is just for detecting infinities.  I **believe** I got those conditions right, but feel free to review them again.

Reply:
Thanks for the answer. I get that this was for the infinity-detection - what confused me was the relation to the statement in the OP that I quoted, which might be better reflected as (because the n=m part already exists):

> Now if z is complex and c=-m, this PR fixes hyp2f1 so that it returns the correctly computed value ~if and only~ _also_ if either a or b equals a negative integer -n where n<m.